src/expr/impl/src/scalar/array_flatten.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Either;
+use risingwave_common::array::{ListRef, ListValue};
+use risingwave_common::types::ScalarRefImpl;
+use risingwave_expr::{Result, function};
+
+/// Flattens a nested array by concatenating the inner arrays into a single array.
+/// Only the outermost level of nesting is removed. For deeper nested arrays, call
+/// `array_flatten` multiple times.
+///
+/// Examples:
+///
+/// ```slt
+/// query T
+/// select array_flatten(array[array[1, 2], array[3, 4]]);
```

Comment:
Just realized `array_flatten(arr)` is equivalent to `genuine_array_concat(VARIADIC arr)`. However:
* This hypothetical `genuine_array_concat` is different from PostgreSQL/RisingWave `array_cat`, which has a wired behavior on >= 2d array due to its weak array typing (`int[][]` is same as `int[]`)
* [#14753](https://github.com/risingwavelabs/risingwave/pull/14753) implements `VARIADIC` as separate exprs but in PostgreSQL there is a single `concat_ws` / `jsonb_extract_path`

Just sharing some related ideas. Nothing to change in this PR.